wallet: reduce unconditional logging during load

mzumsande opened the pull request:
Currently the unconditional log during init with a default wallet happens three times:
```
2025-09-03T19:57:16Z init message: Verifying wallet(s)…
2025-09-03T19:57:16Z Using SQLite Version 3.45.1
2025-09-03T19:57:16Z Using wallet XXX/.bitcoin/regtest
2025-09-03T19:57:16Z Using SQLite Version 3.45.1
2025-09-03T19:57:16Z Using wallet XXX/.bitcoin/regtest
(...)
2025-09-03T19:57:16Z Using SQLite Version 3.45.1
2025-09-03T19:57:16Z Using wallet XXX/.bitcoin/regtest
2025-09-03T19:57:16Z init message: Loading wallet…
```
For non-default wallets it's logged two times.

That seems a bit too much, so just log the SQLite version just one, and remove the log for the full path of the wallet, since it's already clear from other logs which wallet is being loaded.

stickies-v reviewed (commented):
Concept ACK, but I think unconditional logging in the ctor is not the right approach, suggested alternative:

<details>
<summary>git diff on 689a321976</summary>

```diff
diff --git a/src/wallet/init.cpp b/src/wallet/init.cpp
index 79e2c9688d..11d7c9029d 100644
--- a/src/wallet/init.cpp
+++ b/src/wallet/init.cpp
@@ -19,6 +19,7 @@
 #include <util/moneystr.h>
 #include <util/translation.h>
 #include <wallet/coincontrol.h>
+#include <wallet/sqlite.h>
 #include <wallet/wallet.h>
 #include <walletinitinterface.h>
 
@@ -108,6 +109,7 @@ void WalletInit::Construct(NodeContext& node) const
         LogInfo("Wallet disabled!");
         return;
     }
+    LogInfo("Using SQLite Version %s", SQLiteDatabaseVersion());
     auto wallet_loader = node.init->makeWalletLoader(*node.chain);
     node.wallet_loader = wallet_loader.get();
     node.chain_clients.emplace_back(std::move(wallet_loader));
diff --git a/src/wallet/load.cpp b/src/wallet/load.cpp
index 60561b1572..c9c7780893 100644
--- a/src/wallet/load.cpp
+++ b/src/wallet/load.cpp
@@ -85,6 +85,7 @@ bool VerifyWallets(WalletContext& context)
         const auto& wallet_file = wallet.get_str();
         const fs::path path = fsbridge::AbsPathJoin(GetWalletDir(), fs::PathFromString(wallet_file));
 
+        LogInfo("Verifying wallet %s", fs::PathToString(path));
         if (!wallet_paths.insert(path).second) {
             chain.initWarning(strprintf(_("Ignoring duplicate -wallet %s."), wallet_file));
             continue;
diff --git a/src/wallet/sqlite.cpp b/src/wallet/sqlite.cpp
index 0940cbac3b..2e4b9abc1c 100644
--- a/src/wallet/sqlite.cpp
+++ b/src/wallet/sqlite.cpp
@@ -116,8 +116,6 @@ SQLiteDatabase::SQLiteDatabase(const fs::path& dir_path, const fs::path& file_pa
 {
     {
         LOCK(g_sqlite_mutex);
-        LogInfo("Using SQLite Version %s", SQLiteDatabaseVersion());
-        LogInfo("Using wallet %s", fs::PathToString(m_dir_path));
 
         if (++g_sqlite_count == 1) {
             // Setup logging

```
</details>

pablomartin4btc reviewed (commented):
Concept ACK

<details>
<summary>I've managed to reproduce the issue (but I only see the default wallet twice as any named wallets), wallets have to be loaded on startup to verify this problem.</summary>

```
2025-09-06T04:01:48Z init message: Verifying wallet(s)…
2025-09-06T04:01:48Z Using SQLite Version 3.37.2
2025-09-06T04:01:48Z Using wallet /tmp/btc/regtest/wallets
2025-09-06T04:01:48Z Using /16 prefix for IP bucketing
...
2025-09-06T04:01:48Z Using SQLite Version 3.37.2
2025-09-06T04:01:48Z Using wallet /tmp/btc/regtest/wallets
2025-09-06T04:01:48Z init message: Loading wallet…
2025-09-06T04:01:48Z [default wallet] Last client version = 299900
```

</details>

I'd prefer the approach [proposed](https://github.com/bitcoin/bitcoin/pull/33299#pullrequestreview-3189575123) by @stickies-v, which I've tested as I did with this PR's fix, it seems a better place to move the loggings and as a result leaves the code cleaner.

Another suggestion, I know it's not part of the scope of this PR but since you are there in `sqlite.cpp`, if you want you could add a separate commit to replace the `LogPrintf()` calls with `LogInfo()`, something that has been done in different PRs (#32604, #29641) and it would make the code more consistent.

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -135,6 +132,9 @@ SQLiteDatabase::SQLiteDatabase(const fs::path& dir_path, const fs::path& file_pa
         if (ret != SQLITE_OK) {
             throw std::runtime_error(strprintf("SQLiteDatabase: Failed to initialize SQLite: %s\n", sqlite3_errstr(ret)));
         }
+        if (g_sqlite_count == 1) {
```
There's already a similar `count == 1` check above for first time initialization that this log line can be moved into.

mzumsande replied:
I did that at first but changed it on purpose - in case the result of `SQLiteDatabaseVersion()` depended on `sqlite3_initialize` (which it doesn't currently, but doesn't seem completely unreasonable) we'd log the correct thing here too.

achow101 commented:
ACK 862faf3fa7a42238017f8d673b0c656531c688dc

pablomartin4btc reviewed (commented):
tested 862faf3fa7a42238017f8d673b0c656531c688dc

<details>
<summary>This version doesn't work properly and produces the same issue it's trying to fix as every call to <code>MakeWalletDatabase()</code> within conditionals in <code>VerifyWallets()</code> (<i><code>src/wallet/load.cpp</code></i>), ends up also calling the <code>SQLiteDatabase::Cleanup()</code> in the destructor which does <code>--g_sqlite_count</code>.</summary>

```
...
2025-09-19T21:11:13Z Using wallet directory /tmp/btc/regtest/wallets
2025-09-19T21:11:13Z init message: Verifying wallet(s)…
2025-09-19T21:11:13Z Using SQLite Version 3.37.2
2025-09-19T21:11:13Z Using SQLite Version 3.37.2
...
2025-09-19T21:11:13Z init message: Verifying blocks…
2025-09-19T21:11:13Z Block index and chainstate loaded
2025-09-19T21:11:13Z Using SQLite Version 3.37.2
2025-09-19T21:11:13Z init message: Loading wallet…
2025-09-19T21:11:13Z [default wallet] Last client version = 309900
...
```
</details>

<details>
<summary><i>(tiny nit: only if you have to re-touch, there's a typo in commit message body)</i></summary>

<img width="742" height="253" alt="image" src="https://github.com/user-attachments/assets/782be4cd-291c-40b2-8861-f58a45461876" />


</details>

l0rinc commented:
Concept ACK

pablomartin4btc commented on src/wallet/sqlite.cpp:
```diff
@@ -135,6 +132,9 @@ SQLiteDatabase::SQLiteDatabase(const fs::path& dir_path, const fs::path& file_pa
         if (ret != SQLITE_OK) {
             throw std::runtime_error(strprintf("SQLiteDatabase: Failed to initialize SQLite: %s\n", sqlite3_errstr(ret)));
         }
+        if (g_sqlite_count == 1) {
+            LogInfo("Using SQLite Version %s", SQLiteDatabaseVersion());
+        }
```
what about (and no longer need to check `g_sqlite_count`)...
```suggestion
        static bool sqlite_version_logged = false;
        if (!sqlite_version_logged) {
            LogInfo("Using SQLite Version %s", SQLiteDatabaseVersion());
            sqlite_version_logged = true;
        }
```

mzumsande commented:
> This version doesn't work properly and produces the same issue it's trying to fix 

Good catch, must have missed this during testsing! Changed to your suggestion with a `static`.

mzumsande commented:
> This version doesn't work properly and produces the same issue it's trying to fix as every call to MakeWalletDatabase() within conditionals in VerifyWallets() (src/wallet/load.cpp), ends up also calling the SQLiteDatabase::Cleanup() in the destructor which does --g_sqlite_count.

Logging aside, I wonder if it's useful and intended to do all the existing setup including `sqlite3_initialize()`, followed by `sqlite3_shutdown()` multiple times during init for each wallet checked during `VerifyWallet()`? Seems like the intention of `g_sqlite_count` is to do it only once?!

pablomartin4btc commented:
> Logging aside, I wonder if it's useful and intended to do all the existing setup including `sqlite3_initialize()`, followed by `sqlite3_shutdown()` multiple times during init for each wallet checked during `VerifyWallets()`? Seems like the intention of `g_sqlite_count` is to do it only once?!

So, we have one sqlite (/ walletdb/ count) obj instance per wallet (/ wallet file), so during init we check first if there's a default wallet and then also check the wallets in the settings config (what verify does in verify wallets)... we could refactor verify (step 5 in init.cpp) and the load (step 9) together so if the db/ file is opened we keep it... but if there are any that can't be opened that would destroy the instance and decrease the counting. I'd need to investigate if the init steps in between depends on anything that need to be separated (verify first but not keep the wallets opened - step 5 - and load them later - step 9).

Also going back to the approach from @stickies-v I can see now that the dbversion method should be polymorphic with all formats (eg BerkeleyRO), even we only open descriptors at the moment during migration we open the old legacy for read-only purposes.

furszy committed:
wallet, log: reduce unconditional logging during load

The removed statements were logged up to two or three times for each loaded
wallet. The SQLite version only needs to be logged once.
The full wallet path is dropped, since the existing unconditional
logging while loading wallets is sufficient (also reduces anonymization
efforts in case of sharing logs).

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

furszy reviewed (commented):
utACK fc861332b351c9390400054ff74193ce26eb0713

pablomartin4btc reviewed (commented):
Not sure about this, if there are no descriptor wallets during startup (no default, no one in settings), we would be logging the sqlite version but if the user had legacy wallets in settings, that would print a warning (`FAILED_LEGACY_DISABLED` -> `initWarning`) not showing the berkeley version of the DB instantiated (which code is not in `master` either but it doesn't look consistent). If we have a diff DB format in the future we'd need to change this LogDBInfo() (once someone finds it).

I'd prefer first the previous approach or as an alternative move the logic to `MakeDatabase()` in `src/wallet/walletdb.cpp`:
https://github.com/bitcoin/bitcoin/blob/d8fe258cd6105704bf4427eda048dd7085ca516d/src/wallet/walletdb.cpp#L1367-L1373

stickies-v reviewed (approved):
ACK fc861332b351c9390400054ff74193ce26eb0713

nit: Would still prefer doing the logging from `WalletInit::Construct`, `VerifyWallets` feels out of place to me, but no biggie.

<details>
<summary>git diff on fc861332b3</summary>

```diff
diff --git a/src/wallet/init.cpp b/src/wallet/init.cpp
index 79e2c9688d..4013e12967 100644
--- a/src/wallet/init.cpp
+++ b/src/wallet/init.cpp
@@ -20,6 +20,7 @@
 #include <util/translation.h>
 #include <wallet/coincontrol.h>
 #include <wallet/wallet.h>
+#include <wallet/walletdb.h>
 #include <walletinitinterface.h>
 
 using node::NodeContext;
@@ -108,6 +109,7 @@ void WalletInit::Construct(NodeContext& node) const
         LogInfo("Wallet disabled!");
         return;
     }
+    LogDBInfo();
     auto wallet_loader = node.init->makeWalletLoader(*node.chain);
     node.wallet_loader = wallet_loader.get();
     node.chain_clients.emplace_back(std::move(wallet_loader));
diff --git a/src/wallet/load.cpp b/src/wallet/load.cpp
index f362064473..60561b1572 100644
--- a/src/wallet/load.cpp
+++ b/src/wallet/load.cpp
@@ -51,8 +51,6 @@ bool VerifyWallets(WalletContext& context)
     }
 
     LogInfo("Using wallet directory %s", fs::PathToString(GetWalletDir()));
-    // Print general DB information
-    LogDBInfo();
 
     chain.initMessage(_("Verifying wallet(s)…"));
 

```
</details>

furszy commented:
>  If we have a diff DB format in the future we'd need to change this LogDBInfo() (once someone finds it).

That’s the idea. Have a single place where we log all possible DB engines and configurations. The DB engine and its configuration are not specific to any wallet; they are part of the general setup.

> I'd prefer first the previous approach or as an alternative move the logic to MakeDatabase() in src/wallet/walletdb.cpp

That would print the same message multiple times. The goal is to print it only once during init so we know the available DB engine version.

pablomartin4btc commented:
> That’s the idea. Have a single place where we log all possible DB engines and configurations.

At the moment, we use Berkeley DB when we want to migrate a Legacy wallet, the above statement maybe should consider this case.

> That would print the same message multiple times. The goal is to print it only once during init so we know the available DB engine version.

Nop, previous one was 5e130a1e999b8a15d709cb698814c26a85eb5447, but I'm ok to centralize the DB engines logging in one place, still not sure if would be `VerifyWallets` or `WalletInit::Construct`.

pablomartin4btc reviewed (commented):
post-merge ACK https://github.com/bitcoin/bitcoin/commit/fc861332b351c9390400054ff74193ce26eb0713
